llvm/lib/IR/FPAccuracy.cpp

```diff
@@ -0,0 +1,133 @@
+//
+//===----------------------------------------------------------------------===//
+//
+/// @file
+/// This file contains the implementations of functions that map standard
+/// accuracy levels to required accuracy expressed in terms of ulps.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/IR/FPAccuracy.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Intrinsics.h"
+#include <optional>
+
+namespace llvm {
+
+static bool isFPBuiltinIntrinsic(Intrinsic::ID IID) {
+  switch (IID) {
+#define OPERATION(NAME, INTRINSIC) case Intrinsic::INTRINSIC:
+#include "llvm/IR/FPBuiltinOps.def"
+    return true;
+  default:
+    return false;
+  }
+}
+
+static StringRef lookupSyclFloatAccuracy(Intrinsic::ID IID) {
+  switch (IID) {
+#define FP_ACCURACY(INTRINSIC, SYCL_FLOAT_ACCURACY, SDA, CFA, CDA)             \
+  case Intrinsic::INTRINSIC:                                                   \
+    return SYCL_FLOAT_ACCURACY;
+#include "llvm/IR/FPAccuracy.def"
+  default:
+    return StringRef();
+  }
+}
+
+static StringRef lookupSyclDoubleAccuracy(Intrinsic::ID IID) {
+  switch (IID) {
+#define FP_ACCURACY(INTRINSIC, SFA, SYCL_DOUBLE_ACCURACY, CFA, CDA)            \
+  case Intrinsic::INTRINSIC:                                                   \
+    return SYCL_DOUBLE_ACCURACY;
+#include "llvm/IR/FPAccuracy.def"
+  default:
+    return StringRef();
+  }
+}
+
+static StringRef lookupCudaFloatAccuracy(Intrinsic::ID IID) {
+  switch (IID) {
+#define FP_ACCURACY(INTRINSIC, SFA, SDA, CUDA_FLOAT_ACCURACY, CDA)             \
+  case Intrinsic::INTRINSIC:                                                   \
+    return CUDA_FLOAT_ACCURACY;
+#include "llvm/IR/FPAccuracy.def"
+  default:
+    return StringRef();
+  }
+}
+
+static StringRef lookupCudaDoubleAccuracy(Intrinsic::ID IID) {
+  switch (IID) {
+#define FP_ACCURACY(INTRINSIC, SFA, SDA, CFA, CUDA_DOUBLE_ACCURACY)            \
+  case Intrinsic::INTRINSIC:                                                   \
+    return CUDA_DOUBLE_ACCURACY;
+#include "llvm/IR/FPAccuracy.def"
+  default:
+    return StringRef();
+  }
+}
+
+StringRef fp::getAccuracyForFPBuiltin(Intrinsic::ID IID,
+                                                 const Type *Ty,
+                                                 fp::FPAccuracy AccuracyLevel) {
+  assert(isFPBuiltinIntrinsic(IID) && "Invalid intrinsic ID for FPAccuracy");
+
+  assert(Ty->isFPOrFPVectorTy() && "Invalid type for FPAccuracy");
+
+  // Vector fpbuiltins have the same accuracy requirements as the corresponding
+  // scalar operation.
+  if (const auto *VecTy = dyn_cast<VectorType>(Ty))
+    Ty = VecTy->getElementType();
+
+  // This will probably change at some point.
+  assert((Ty->isFloatTy() || Ty->isDoubleTy()) &&
+         "Invalid type for FPAccuracy");
+
+  // High and medium accuracy have the same requirement for all functions
```

Comment:
Is there documentation for how these values are selected? May be a link here can be handy?

Reply:
For high, medium, and low, the accuracy requirements are copied from the icc compiler's -imf-precision=[high|medium|low] option. To those who have never used icc, these values are essentially arbitrary, but we've found them to be good choices for users.